src/orion/client/experiment.py

```diff
@@ -622,26 +621,34 @@ def workon(self, fct, max_trials=infinity, n_jobs=None, **kwargs):
         """
         self._check_if_executable()
 
-        with Parallel(n_jobs=n_jobs, verbose=100) as parallel:
+        # reduce chance of race condition for trials creation
+        if self.max_trials > max_trials:
+            self._experiment.max_trials = max_trials
+            self._experiment.algorithms.algorithm.max_trials = max_trials
+
+        with Parallel(n_jobs=n_workers, verbose=100) as parallel:
             trials = parallel(
-                delayed(self._optimize)(fct, kwargs) for _ in range(max_trials)
+                delayed(self._optimize)(fct, max_trials, kwargs)
+                for _ in range(n_workers)
             )
 
         return sum(trials)
 
-    def _optimize(self, fct, *args, **kwargs):
-
+    def _optimize(self, fct, max_trials, *args, **kwargs):
         # this is required for process-based or remote backend
         setup_storage(storage=self.storage_config)
 
-        trial = self.suggest()
-        if trial is None:
-            log.warning("Algorithm could not sample new points")
-            return 0
-        kwargs.update(flatten(trial.params))
-        results = fct(**unflatten(kwargs))
-        self.observe(trial, results=results)
-        return 1
+        trials = 0
+        while not self.is_done and trials < max_trials:
```

Comment:
`trials < max_trials` should never be reached right? If `max_trials` is smaller than `exp.max_trials`, than `exp.max_trials` is set to `max_trials`. So here `max_trials` can only be equal or larger than `exp.max_trials`. Since `exp.is_done` depends on `num_trials < exp.max_trials`, then `exp.is_done` will always be True whenever `trials < max_trials` is True.

Reply:
It is suppose to be, and this is the same code as the old loop.  `self.is_done` should already cover `trials < max_trials`.

Reply:
Yeah the difference now is that exp.max_trial is set to `max_trial` if `max_trial < exp.max_trial`. So I think we could remove `trials < max_trials` in the while clause.